### ISSUE TITLE
Sign2CodeTrans generates invalid code for an array argument

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#2768]
Simon has reported that `Sign2CodeTrans` produces invalid code when applied to a SIGN that has an array as argument. One of the examples of this is from src/OCE/SBC/sbcblk_algo_andreas.F90 in NEMO from which I've constructed the following:
```
    program test_prog
      integer, parameter :: wp = kind(1.0d0)
      integer, parameter, dimension(0:4) :: A2D = (/1, 2, 3, 4, 5/)
      REAL(wp), DIMENSION(A2D(0)) :: ztmp1
      ztmp1 = 0.0
      ztmp1 = SIGN( MAX(ABS(ztmp1),1.E-6_wp), ztmp1 )
    end program test_prog'''
```

**Comment by arporter** [comment]
The code PSyclone currently generates for this is:
```
  tmp_abs = MAX(ABS(ztmp1), 1.e-6_wp)
  if (tmp_abs > 0.0) then
    res_abs = tmp_abs
  else
    res_abs = tmp_abs * -1.0
  end if
```
and the problem is that `tmp_abs` is declared as a scalar but `MAX(ABS(ztmp1), 1.0E-6)` is an array.

**Comment by arporter** [comment]
Alternatively, could we apply `Reference2ArrayRangeTrans` to
```
  ztmp1 = SIGN( MAX(ABS(ztmp1),1.E-6_wp), ztmp1 )
```
to get:
```
  ztmp1(:) = SIGN( MAX(ABS(ztmp1(:)),1.E-6_wp), ztmp1(:) )
```
and then `ArrayAssignment2LoopsTrans` to get:
```
  do i = 1, ...
     ztmp1(i) = SIGN(MAX(ABS(ztmp1(i)), 1.E-6_wp), ztmp1(i))
  end do
```
At which point, the Sign2CodeTransformation should work OK?

**Comment by arporter** [comment]
The transformation itself had a comment saying we assume that the argument is a real scalar because we can't query the datatype (with a reference to an old issue). I've closed that old issue (#658) and am proceeding to do better. However, I immediately hit the fact that we don't know what the return type of MAX or ABS is and so the datatype of the argument to SIGN comes back as UnresolvedType. We hit this situation because we can't assume that MAX and ABS are definitely the eponymous Fortran intrinsics.

**Comment by arporter** [comment]
It turns out that ABS2CodeTrans made the same assumption about operating on a scalar, real argument so I've generalised my fix (validation really) so that it is used there too.
